northstar/src/npk/npk.rs

```diff
@@ -635,58 +644,115 @@ fn sign_npk(key: &Path, fsimg: &Path, manifest: &Manifest) -> Result<String, Err
 
 /// Returns a temporary file with all the pseudo file definitions
 fn gen_pseudo_files(manifest: &Manifest) -> Result<NamedTempFile, Error> {
-    let pseudo_file_entries = NamedTempFile::new().map_err(|error| Error::Io {
+    let mut pseudo_file_entries = NamedTempFile::new().map_err(|error| Error::Io {
         context: "Failed to create temporary file".to_string(),
         error,
     })?;
-    let file = pseudo_file_entries.as_file();
+    let file = pseudo_file_entries.as_file_mut();
     let uid = manifest.uid;
     let gid = manifest.gid;
 
     fn add_directory(
-        mut file: &fs::File,
-        directory: &Path,
+        file: &mut fs::File,
+        dir: &Path,
         mode: u16,
         uid: u16,
         gid: u16,
     ) -> Result<(), Error> {
-        writeln!(file, "{} d {} {} {}", directory.display(), mode, uid, gid).map_err(|e| {
-            Error::Io {
-                context: format!("Failed to write entry {} to temp file", directory.display()),
-                error: e,
-            }
-        })
-    }
-
-    if manifest.init.is_some() {
-        // The default is to have at least a minimal /dev mount
-        add_directory(file, Path::new("/dev"), 444, uid, gid)?;
-        add_directory(file, Path::new("/proc"), 444, uid, gid)?;
+        // Each directory level needs to be passed to mksquashfs e.g:
+        // /dev d 755 x x x
+        // /dev/block d 755 x x x
+        dir.iter()
+            .skip(1)
+            .try_fold(PathBuf::from("/"), |mut p, dir| {
+                p.push(dir);
+                writeln!(file, "{} d {} {} {}", p.display(), mode, uid, gid).map_err(|e| {
+                    Error::Io {
+                        context: format!("Failed to write entry {} to temp file", p.display()),
+                        error: e,
+                    }
+                })?;
+                Ok(p)
+            })
+            .map(drop)
```

Comment:
nit: This is a pretty cool iterator expression, but maybe in this particular case a good ol' for loop could be a bit more readable?

```rust
let mut p = PathBuf::from("/");

for d in dir.iter().skip(1) {
    p.push(d);
    writeln!(file, "{} d {} {} {}", p.display(), mode, uid, gid).map_err(|e| {
        Error::Io {
            context: format!("Failed to write entry {} to temp file", p.display()),
            error: e,
        }
    })?;
}
```

Reply:
Also it looks like this function is big enough to be moved to the outer scope and ideally have it documented and unit tested?

Reply:
Thanks for the hint!. I refactored the whole function. The writes on line were just too noisy.